### PR TITLE
fix issue #12281

### PR DESCRIPTION
**Pull request by dalanlan**
Fix issue #12281 
Reject the client `pull` request for special case `scratch`.
Signed-off-by: Simei He hesimei@zju.edu.cn

### CODE REVIEW

**Comment by dalanlan** [comment]
@thaJeztah  @dmcgowan

**Comment by dmcgowan** [comment]
Setting this to design review to discuss whether we want to do this.  The code looks good though.

**Comment by mmdriley** [comment]
Really seems like this should use validateRepositoryName rather than duplicate that logic.

https://github.com/docker/docker/blob/fe53c277857ba77caf9189b1d6d095835c1b792c/registry/config.go#L250

**Comment by dalanlan** [comment]
Thx @mmdriley  
I thought about that before, however considering the function you mentioned is called in multiple cases `build`,`commit`,`import`,etc, i wouldnt regard it as an appropriate solution to go that deep.

**Comment by mmdriley** [comment]
I suggest using validateRepositoryName precisely *because* it's so commonly used. That's no accident -- it was [a lot of work](https://github.com/docker/docker/pull/8456) to centralize repository name parsing and validation.

File: graph/pull.go [hunk 1]
```diff
@@ -40,6 +59,10 @@ func (s *TagStore) Pull(image string, tag string, imagePullConfig *ImagePullConf
 		return err
 	}
 
+	if err := validateRepoNotScratch(repoInfo.RemoteName); err != nil {
```
Comment (jlhawn) [inline graph/pull.go]: please use `validateRepoName(repoInfo.LocalName)` instead. Defining a separate `validateRepoNotScratch()` function is not necessary.

**Comment by jlhawn** [comment]
needs rebase

**Comment by jlhawn** [comment]
@dalanlan I'm not sure if you are available to update this PR, so I've quickly submitted another here: #12655

**Comment by jfrazelle** [comment]
closing as there is a new pr https://github.com/docker/docker/pull/12655

**Pull request closed without merging**